services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/SemanticDetector.java

```diff
@@ -0,0 +1,438 @@
+package com.linkedin.venice.controller.kafka.protocol.serializer;
+
+import com.linkedin.venice.exceptions.VeniceProtocolException;
+import com.linkedin.venice.utils.AvroSchemaUtils;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+
+/**
+ * SemanticDetector is used to detect the semantic changes between two schemas.
+ * It traverses the object based on the current schema and target schema.
+ * When we detect new semantic change, we compare object with default value.
+ * If the object is NOT default value, we throw @code{VeniceProtocolException}.
+ *
+ * Semantic changes include:
+ * 1. Changing the type of a field
+ * 2. Adding a new field
+ * 3. Changing the fixed size of a fixed type
+ * 4. Adding a new enum value
+ */
+public class SemanticDetector {
+  /**
+   * Traverse the object based on the current schema and target schema. Stop traversing if found a semantic change &
+   * using not default value.
+   *
+   * @param object        the object containing values
+   * @param currentSchema the object schema
+   * @param targetSchema  the target schema we want to compare with
+   * @param name          The name of the object from the parent object
+   * @param defaultValue  The default value of the object
+   * @throws VeniceProtocolException if there is a semantic change between the two schemas & value from
+   *                                 @code{@param object} is non-default value.
+   */
+  public static void traverseAndValidate(
+      Object object,
+      Schema currentSchema,
+      Schema targetSchema,
+      String name,
+      Object defaultValue) {
+    // If target schema is null, this is a new field and we need to validate the object against the current schema.
+    if (targetSchema == null) {
+      compareObjectToDefaultValue(object, currentSchema, name, defaultValue);
+      return;
+    }
+
+    // If the current schema and target schema do not have the same type, we return true. This is a semantic change.
+    if (currentSchema.getType() != targetSchema.getType()) {
+      throw new VeniceProtocolException(
+          String.format(
+              "Field %s: Type %s is not the same as %s",
+              formatFieldName(name),
+              currentSchema.getType(),
+              targetSchema.getType()));
+    }
+
+    // If the current schema and target schema are the same, we do not need to traverse the object.
+    // @code{compareSchemaIgnoreFieldOrder} method is used to compare the two schemas (but cannot take FIXED type)
```

Comment:
Please add a more context here to explain why we need to do more checks for FIXED type even if both current and target schema has FIXED type.